impl/src/main/java/org/jboss/weld/event/TransactionalObserverNotifier.java

```diff
@@ -73,8 +78,32 @@ protected TransactionalObserverNotifier(String contextId, TypeSafeObserverResolv
             // Transaction is not active - no deferred notifications
             notifySyncObservers(observers, event, metadata);
         } else {
-            for (ObserverMethod<? super T> observer : observers) {
-                deferNotification(event, metadata, observer);
+            try {
+                for (ObserverMethod<? super T> observer : observers) {
+                    deferNotification(event, metadata, observer);
+                }
+            } catch (Exception e) {
+                if (e.getCause() instanceof RollbackException || e.getCause() instanceof IllegalStateException) {
+
+                    List<ObserverMethod<? super T>> filteredObservers = new ArrayList<ObserverMethod<? super T>>();
+                    filteredObservers.addAll(observers);
+                    for (ObserverMethod<? super T> observer : observers) {
+                        if (observer.getTransactionPhase().equals(TransactionPhase.AFTER_SUCCESS)) {
+                            filteredObservers.remove(observer);
+                        }
+                    }
+
+                    Collections.sort(filteredObservers, new Comparator<ObserverMethod<? super T>>() {
+                        @Override
+                        public int compare(ObserverMethod<? super T> o1, ObserverMethod<? super T> o2) {
+                            return o1.getTransactionPhase().compareTo(o2.getTransactionPhase());
```

Comment:
This will compare enums by ordinals. Although it's unlikely that ordering of `TransactionPhase` values will change it might be better to compare by values toString()...?